src/OpenSpaceToolkit/Astrodynamics/Flight/Profile/Model/Tabulated.cpp

```diff
@@ -255,98 +290,62 @@ bool Tabulated::operator!=(const Model& aModel) const
     return !((*this) == aModel);
 }
 
-Pair<const State*, const State*> Tabulated::accessStateRangeAt(const Instant& anInstant) const
+void Tabulated::setMembers(const Array<State>& aStateArray, const Interpolator::Type& anInterpolatorType)
 {
-    State const* previousStatePtr = nullptr;
-    State const* nextStatePtr = nullptr;
-
-    while (true)  // To be improved
+    if (aStateArray.getSize() < 2)
     {
-        Unpack(previousStatePtr, nextStatePtr) = this->accessStateRangeAtIndex(stateIndex_);  // Check index cache
+        return;
+    }
+
+    const State& firstState = aStateArray.accessFirst();
+
+    Array<Shared<const CoordinateSubset>> coordinateSubsets = firstState.getCoordinateSubsets();
+
+    // ensure that the last coordinate subset is the attitude quaternion
+    coordinateSubsets.remove(AttitudeQuaternion::Default());
+    coordinateSubsets.add(AttitudeQuaternion::Default());
 
-        if ((previousStatePtr != nullptr) && (nextStatePtr != nullptr))
+    stateBuilder_ = StateBuilder(aStateArray[0].accessFrame(), coordinateSubsets);
+
+    Array<State> stateArray = aStateArray;
+
+    std::sort(
+        stateArray.begin(),
+        stateArray.end(),
+        [](const auto& lhs, const auto& rhs)
         {
-            if ((previousStatePtr->accessInstant() <= anInstant) && (anInstant <= nextStatePtr->accessInstant()))
-            {
-                if (previousStatePtr->accessInstant() == anInstant)
-                {
-                    return {nullptr, previousStatePtr};
-                }
-                else if (nextStatePtr->accessInstant() == anInstant)
-                {
-                    return {nextStatePtr, nullptr};
-                }
-
-                return {previousStatePtr, nextStatePtr};
-            }
-            else
-            {
-                if (anInstant < previousStatePtr->accessInstant())
-                {
-                    if (stateIndex_ > 0)
-                    {
-                        stateIndex_--;
-                    }
-                    else
-                    {
-                        break;
-                    }
-                }
-                else
-                {
-                    if ((states_.getSize() > 0) && (stateIndex_ < (states_.getSize() - 1)))
-                    {
-                        stateIndex_++;
-                    }
-                    else
-                    {
-                        break;
-                    }
-                }
-            }
+            return lhs.getInstant() < rhs.getInstant();
         }
-        else if (previousStatePtr != nullptr)
+    );
+
+    stateArray_ = stateArray;
+
+    VectorXd timestamps(stateArray.getSize());
+    MatrixXd coordinates(stateArray.getSize(), firstState.getSize() - 4);  // Exclude quaternion
```

Comment:
could we instead do:
```suggestion
    MatrixXd coordinates(stateArray.getSize(), firstState.getSize() - AttitudeQuaternion::Default().getSize());  // Exclude quaternion
```
?